doc/coq.rst

```diff
@@ -0,0 +1,383 @@
+
+.. _limitation-mlpack:
+
+- A ``foo.mlpack`` file must the present in directories of locally defined
+  plugins for things to work. This is a limitation of ``coqdep``.
+
+- Building a theory and a plugin in the same directory can lead to issues with
+  the presence of the META file. We recommend the following:
+  - A separate directory for the files of each :ref:`coq-theory` stanza defined.
+  - A separate directory for source files of a plugin.
+
+.. _coq-lang:
+
+Coq Language Version
+~~~~~~~~~~~~~~~~~~~~
+
+The Coq lang can be modified by adding the following to a :ref:`dune-project`
+file:
+
+.. code:: scheme
+
+    (using coq 0.3)
+
+The supported Coq language versions (not the version of Coq) are:
+
+- ``0.1``: Basic Coq theory support.
+- ``0.2``: Support for the ``theories`` field and composition of theories in the
+  same scope,
+- ``0.3``: Support for ``(mode native)`` requires Coq >= 8.10 (and Dune >= 2.9
+  for Coq >= 8.14).
+
+Guarantees with respect to stability are not yet provided. However, as the
+development of features progresses, we hope to reach ``1.0`` soon. The ``1.0``
+version of Coq lang will commit to a stable set of functionality. All the
+features below are expected to reach ``1.0`` unchanged or minimally modified.
+
+.. _coq-extraction:
+
+coq.extraction
+--------------
+
+Coq may be instructed to *extract* OCaml sources as part of the compilation
+process by using the ``coq.extraction`` stanza:
+
+.. code:: scheme
+
+   (coq.extraction
+    (prelude <name>)
+    (extracted_modules <names>)
+    <optional-fields>)
+
+- ``(prelude <name>)`` refers to the Coq source that contains the extraction
+  commands.
+
+- ``(extracted_modules <names>)`` is an exhaustive list of OCaml modules
+  extracted.
+
+- ``<optional-fields>`` are ``flags``, ``theories``, and ``libraries``. All of
+  these fields have the same meaning as in the ``coq.theory`` stanza.
+
+The extracted sources can then be used in ``executable`` or ``library`` stanzas
+as any other sources.
+
+Note that the sources are extracted to the directory where the ``prelude`` file
+lives. Thus the common placement for the ``OCaml`` stanzas is in the same
+:ref:`dune file<dune-files>`.
+
+**Warning**: using Coq's ``Cd`` command to work around problems with the output
+directory is not allowed when using extraction from Dune. Moreover the ``Cd``
+command has been deprecated in Coq 8.12.
+
+.. _coq-pp:
+
+coq.pp
+------
+
+Authors of Coq plugins often need to write ``.mlg`` files to extend the Coq
+grammar. Such files are preprocessed with the ``coqpp`` binary. To help plugin
+authors avoid writing boilerplate, we provide a ``(coqpp ...)`` stanza:
+
+.. code:: scheme
+
+    (coq.pp (modules <mlg_list>))
+
+which, for each ``g_mod`` in ``<mlg_list>``, is equivalent to the following
+rule:
+
+.. code:: lisp
+
+    (rule
+     (targets g_mod.ml)
+     (deps (:mlg-file g_mod.mlg))
+     (action (run coqpp %{mlg-file})))
+
+Examples of Coq Projects
```

Comment:
You should mention the examples above and hyperlink to them.  The preceding text will make more sense if read with the examples.

Reply:
I've added some more hyperlinks.